cpp/src/parquet/size_statistics.cc

```diff
@@ -0,0 +1,266 @@
+}
+
+class SizeStatisticsBuilder::SizeStatisticsBuilderImpl {
+ public:
+  SizeStatisticsBuilderImpl(const ColumnDescriptor* descr)
+      : rep_level_histogram_(descr->max_repetition_level() + 1, 0),
+        def_level_histogram_(descr->max_definition_level() + 1, 0) {
+    if (descr->physical_type() == Type::BYTE_ARRAY) {
+      unencoded_byte_array_data_bytes_ = 0;
+    }
+  }
+
+  void WriteRepetitionLevels(int64_t num_levels, const int16_t* rep_levels) {
+    for (int64_t i = 0; i < num_levels; ++i) {
+      ARROW_DCHECK_LT(rep_levels[i], static_cast<int16_t>(rep_level_histogram_.size()));
+      rep_level_histogram_[rep_levels[i]]++;
+    }
+  }
+
+  void WriteDefinitionLevels(int64_t num_levels, const int16_t* def_levels) {
+    for (int64_t i = 0; i < num_levels; ++i) {
+      ARROW_DCHECK_LT(def_levels[i], static_cast<int16_t>(def_level_histogram_.size()));
+      def_level_histogram_[def_levels[i]]++;
+    }
+  }
+
+  void WriteRepetitionLevel(int64_t num_levels, int16_t rep_level) {
+    ARROW_DCHECK_LT(rep_level, static_cast<int16_t>(rep_level_histogram_.size()));
+    rep_level_histogram_[rep_level] += num_levels;
+  }
+
+  void WriteDefinitionLevel(int64_t num_levels, int16_t def_level) {
+    ARROW_DCHECK_LT(def_level, static_cast<int16_t>(def_level_histogram_.size()));
+    def_level_histogram_[def_level] += num_levels;
+  }
+
+  void WriteValuesSpaced(const ByteArray* values, const uint8_t* valid_bits,
+                         int64_t valid_bits_offset, int64_t num_spaced_values) {
+    int64_t total_bytes = 0;
+    ::arrow::internal::VisitSetBitRunsVoid(valid_bits, valid_bits_offset,
+                                           num_spaced_values,
+                                           [&](int64_t pos, int64_t length) {
+                                             for (int64_t i = 0; i < length; i++) {
+                                               // Don't bother to check unlikely overflow.
+                                               total_bytes += values[i + pos].len;
+                                             }
+                                           });
+    IncrementUnencodedByteArrayDataBytes(total_bytes);
+  }
+
+  void WriteValues(const ByteArray* values, int64_t num_values) {
+    int64_t total_bytes = 0;
+    std::for_each(values, values + num_values,
+                  [&](const ByteArray& value) { total_bytes += values->len; });
+    IncrementUnencodedByteArrayDataBytes(total_bytes);
+  }
+
+  void WriteValues(const ::arrow::Array& values) {
+    int64_t total_bytes = 0;
+    const auto valid_func = [&](ByteArray val) { total_bytes += val.len; };
+    const auto null_func = [&]() {};
+
+    if (::arrow::is_binary_like(values.type_id())) {
+      ::arrow::VisitArraySpanInline<::arrow::BinaryType>(
+          *values.data(), std::move(valid_func), std::move(null_func));
+    } else if (::arrow::is_large_binary_like(values.type_id())) {
```

Comment:
i might have read to quickly above but how does this account for `dictionary<string>` arrays?

Reply:
Same to my comment at https://github.com/apache/arrow/pull/40594#discussion_r1649733580. The `referenced_dictionary` simply puts all referenced string entries into a `arrow::StringArray` so we can directly consume it here.